metadata-builders/src/main/java/com/linkedin/metadata/builders/search/MLModelIndexBuilder.java

```diff
@@ -0,0 +1,156 @@
+
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class MLModelIndexBuilder extends BaseIndexBuilder<MLModelDocument> {
+
+    public MLModelIndexBuilder() {
+        super(Collections.singletonList(MLModelSnapshot.class), MLModelDocument.class);
+    }
+
+    @Nonnull
+    private static String buildBrowsePath(@Nonnull MLModelUrn urn) {
+        return ("/" + urn.getOriginEntity() + "/"  + urn.getPlatformEntity().getPlatformNameEntity() + "/" + urn.getMlModelNameEntity())
+            .replace('.', '/').toLowerCase();
+    }
+
+    /**
+     * Given model urn, this returns a {@link MLModelDocument} model that has urn, dataset name, platform and origin fields set
+     *
+     * @param urn {@link MLModelUrn} that needs to be set
+     * @return {@link MLModelDocument} model with relevant fields set that are extracted from the urn
+     */
+    @Nonnull
+    private static MLModelDocument setUrnDerivedFields(@Nonnull MLModelUrn urn) {
+        return new MLModelDocument()
+            .setName(urn.getMlModelNameEntity())
+            .setOrigin(urn.getOriginEntity())
+            .setPlatform(urn.getPlatformEntity().getPlatformNameEntity())
+            .setUrn(urn);
+    }
+
+    @Nonnull
+    private List<MLModelDocument> getDocumentsToUpdateFromSnapshotType(@Nonnull MLModelSnapshot mlModelSnapshot) {
+        final MLModelUrn urn = mlModelSnapshot.getUrn();
+        final List<MLModelDocument> documents =  mlModelSnapshot.getAspects().stream().map(aspect -> {
+            if (aspect.isDeprecation()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getDeprecation());
+            } else if (aspect.isEvaluationData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getEvaluationData());
+            } else if (aspect.isMLModelProperties()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMLModelProperties());
+            } else if (aspect.isOwnership()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getOwnership());
+            } else if (aspect.isStatus()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getStatus());
+            } else if (aspect.isTrainingData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getTrainingData());
+            }
+            return null;
+        }).filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        documents.add(setUrnDerivedFields(urn));
+        return documents;
+    }
+
+    @Nonnull
+    private MLModelDocument getDocumentToUpdateFromAspect(MLModelUrn urn, Deprecation deprecation) {
+        return new MLModelDocument()
+            .setActive(!deprecation.isDeprecated());
+    }
+
+    @Nonnull
+    private MLModelDocument getDocumentToUpdateFromAspect(MLModelUrn urn, EvaluationData evaluationData) {
+        final MLModelDocument doc = new MLModelDocument();
+
+        if (evaluationData.hasEvaluationData()) {
```

Comment:
more of a generic comment. `hasX()` doesn't guarantee that `getX()` will not be null. I suggest we replace this with `(evaluationData.getEvaluationData() != null)`. If you can make similar changes in other parts of the code that will be great! Related discussion https://github.com/linkedin/datahub/pull/1950#discussion_r509753102